### PR TITLE
Add projsync utility

### LINKED ISSUE #1750: Add a projsync utility
Idea mentioned at bootoom of https://lists.osgeo.org/pipermail/proj/2019-November/009032.html

### PR DESCRIPTION
**Pull request by rouault**
Fixes #1750

### CODE REVIEW

**Comment by hobu** [comment]
Why do we have `projinfo` and then `proj_download`? What discriminates the underscore separation? GDAL's behavior on this topic is so arbitrary that it requires lookup just about every time you use the tools. Let's not repeat that with PROJ.

**Comment by hobu** [comment]
I wonder if this should be called some variant of `sync`. It isn't a generic download tool. We could expand it to also include fetching latest DBs too, which would provide a lot of convenience.

**Review by kbevers — COMMENTED** [review]
Looks good to me. This is a nice utility to add the package. 

I wonder about the tests of the application, how resilient are they to changes in PROJ-data/the CDN? Do we need to update the tests everytime a change is made on the CDN or will this still work unchanged?

**Comment by rouault** [comment]
@hobu I was indeed midly satisfied with the name. projdownload seemed a bit stodgy to me, but projsync would be good. I'll go for it, if you and @kbevers is happy with that

Regarding synchronizing proj.db, this is a hard topic as I discussed a number of times. PROJ is highly sensitive to the content of proj.db. So if we were to do it, we would likely to have proj_7.0.0.db files compatible of PROJ 7.0.0 binary.

**Comment by rouault** [comment]
> how resilient are they to changes in PROJ-data/the CDN?

They should be pretty resilient, unless we remove some files / source agencies.

**Comment by hobu** [comment]
> So if we were to do it, we would likely to have proj_7.0.0.db files compatible of PROJ 7.0.0 binary.

My thought was to update if a new version of EPSG came out. Maybe it isn't worth the trouble.

**Comment by kbevers** [comment]
> I'll go for it, if you and @kbevers is happy with that

Sounds good to me.

> My thought was to update if a new version of EPSG came out. Maybe it isn't worth the trouble.

It seems that EPSG doesn't really follow semantic versioning, so an increase in the patch-number can add a new operation or some other thing that requires new code. #1869 is a good example of that. It would be great if we in the future could decouple the EPSG-registry from specific PROJ releases and update the database when new updates appear in the EPSG registry but we're not there yet.

File: docs/source/apps/projsync.rst [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+projsync
+================================================================================
+
+.. Index:: projsync
+
+.. only:: html
+
+    .. versionadded:: 7.0.0
+
+    Tool for downloading resource files
```
Comment (hobu) [inline docs/source/apps/projsync.rst]: Tool for synchronizing PROJ datum and transformation support data.

File: docs/source/apps/projsync.rst [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    Restrict resource files to be downloaded to those whose name property
+    contains the NAME value. Specifying ``?`` as NAME will list all possible values.
+
+.. option:: --all
+
+    Ask to download all files.
+
+.. option:: --exclude-world-coverage
+
+    Exclude files which have world coverage.
```
Comment (hobu) [inline docs/source/apps/projsync.rst:105]: What box defines world coverage?

File: docs/source/apps/projsync.rst [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    See also
+    ********
+
+    **cs2cs(1)**, **cct(1)**, **geod(1)**, **gie(1)**, **proj(1)**, **projinfo(1)**
+
+    Bugs
+    ****
+
+    A list of know bugs can be found at https://github.com/OSGeo/PROJ/issues
+    where new bug reports can be submitted to.
```
Comment (hobu) [inline docs/source/apps/projsync.rst:153]: It would be helpful but maybe confusing to also link to the data repo

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -15,6 +15,8 @@ option(BUILD_PROJ
   "Build proj (cartographic projection tool)" ON)
 option(BUILD_PROJINFO
   "Build projinfo (SRS and coordinate operation metadata/query tool)" ON)
+option(BUILD_PROJSYNC
+  "Build projsync (Data downloader tool)" ON)
```
Comment (hobu) [inline src/CMakeLists.txt]: Build projsync (synchronize transformation support data)

File: src/apps/projsync.cpp [hunk 5]
```diff
@@ -0,0 +1,556 @@
...
+                                grid_bbox[1] = std::min(grid_bbox[1], bbox[1]);
+                                grid_bbox[2] = std::max(grid_bbox[2], bbox[2]);
+                                grid_bbox[3] = std::max(grid_bbox[3], bbox[3]);
+                            }
+                        }
+                    } else {
+                        grid_bbox = get_bbox(j_coordinates);
+                    }
+                    if (grid_bbox.size() != 4) {
+                        break;
```
Comment (hobu) [inline src/apps/projsync.cpp:440]: Are there boxes with Zs too?

File: src/apps/projsync.cpp [hunk 6]
```diff
@@ -0,0 +1,556 @@
...
+                        grid_e = std::numeric_limits<double>::max();
+                    } else if (grid_e > 180 && queried_west < -180) {
+                        grid_w -= 360;
+                        grid_e -= 360;
+                    }
+                    if (queryAll) {
+                        matchBbox = true;
+                        break;
+                    }
+                    // std::cout << name << " " << grid_w  << " " << grid_s << "
```
Comment (hobu) [inline src/apps/projsync.cpp]: cruft

File: src/bin_projsync.cmake [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+add_executable(bin_projsync ${PROJSYNC_SRC})
+set_target_properties(bin_projsync
+  PROPERTIES
+  OUTPUT_NAME projsync)
+target_link_libraries(bin_projsync ${PROJ_LIBRARIES})
+target_compile_options(bin_projsync PRIVATE ${PROJ_CXX_WARN_FLAGS})
+install(TARGETS bin_projsync
+  RUNTIME DESTINATION ${BINDIR})
+
+if(MSVC AND BUILD_LIBPROJ_SHARED)
```
Comment (hobu) [inline src/bin_projsync.cmake:14]: Use the standard BUILD_SHARED_LIBS cmake variable instead

File: src/filemanager.cpp [hunk 8]
```diff
@@ -1178,6 +1178,8 @@ static void CreateDirectoryRecursively(PJ_CONTEXT *ctx,
 
 std::string pj_context_get_user_writable_directory(PJ_CONTEXT *ctx,
                                                    bool create) {
+    if (!ctx)
```
Comment (hobu) [inline src/filemanager.cpp:1181]: Was this missing before?

File: src/filemanager.hpp [hunk 9]
```diff
@@ -53,8 +53,8 @@ class FileManager {
 
   public:
     // "Low-level" interface.
-    static std::unique_ptr<File> open(PJ_CONTEXT *ctx, const char *filename,
-                                      FileAccess access);
+    static PROJ_DLL std::unique_ptr<File>
```
Comment (hobu) [inline src/filemanager.hpp:56]: thread synchronization concerns with this?

**Review by hobu — APPROVED** [review]

File: docs/source/apps/projsync.rst (hunk 1 above)
Comment (rouault) [reply docs/source/apps/projsync.rst]: done

File: docs/source/apps/projsync.rst (hunk 2 above)
Comment (rouault) [reply docs/source/apps/projsync.rst:105]: "grid_e - grid_w > 359 && grid_n - grid_s > 179", to exclude EGM files

File: docs/source/apps/projsync.rst (hunk 3 above)
Comment (rouault) [reply docs/source/apps/projsync.rst:153]: Adding "Bugs specific to resource files should be submitted to
    https://github.com/OSGeo/PROJ-data/issues"

File: src/CMakeLists.txt (hunk 4 above)
Comment (rouault) [reply src/CMakeLists.txt]: done

File: src/apps/projsync.cpp (hunk 5 above)
Comment (rouault) [reply src/apps/projsync.cpp:440]: no. sanity check only

File: src/apps/projsync.cpp (hunk 6 above)
Comment (rouault) [reply src/apps/projsync.cpp]: removed

File: src/bin_projsync.cmake (hunk 7 above)
Comment (rouault) [reply src/bin_projsync.cmake:14]: All other similar files, for better or worse (likely the later), use BUILD_LIBPROJ_SHARED, so I'll keep that for consistency.

File: src/filemanager.cpp (hunk 8 above)
Comment (rouault) [reply src/filemanager.cpp:1181]: likely

File: src/filemanager.hpp (hunk 9 above)
Comment (rouault) [reply src/filemanager.hpp:56]: I don't see any thread related issue here

**Pull request merged**